indexer/trie_builder.hpp

```diff
@@ -150,33 +188,50 @@ void PopNodes(TSink & sink, TNodes & nodes, int nodesToPop)
     TNodeInfo & node = nodes.back();
     TNodeInfo & prevNode = nodes[nodes.size() - 2];
 
-    if (node.m_valueList.empty() && node.m_children.size() <= 1)
+    if (node.m_temporaryValueList.empty() && node.m_children.size() <= 1)
     {
       ASSERT_EQUAL(node.m_children.size(), 1, ());
       ChildInfo & child = node.m_children[0];
-      prevNode.m_children.push_back(ChildInfo(child.m_isLeaf, child.m_size, node.m_char));
-      prevNode.m_children.back().m_edge.append(child.m_edge.begin(), child.m_edge.end());
+      prevNode.m_children.emplace_back(child.m_isLeaf, child.m_size, node.m_char);
+      auto & prevChild = prevNode.m_children.back();
+      prevChild.m_edge.insert(prevChild.m_edge.end(), child.m_edge.begin(), child.m_edge.end());
     }
     else
     {
-      WriteNodeReverse(sink, node.m_char, node);
-      prevNode.m_children.push_back(ChildInfo(node.m_children.empty(),
-                                              static_cast<uint32_t>(sink.Pos() - node.m_begPos),
-                                              node.m_char));
+      WriteNodeReverse(sink, serializer, node.m_char, node);
+      prevNode.m_children.emplace_back(
+          node.m_children.empty(), static_cast<uint32_t>(sink.Pos() - node.m_begPos), node.m_char);
     }
 
     nodes.pop_back();
   }
 }
 
-template <typename TSink, typename TIter, typename TValueList>
-void Build(TSink & sink, TIter const beg, TIter const end)
+template <typename TNodeInfo, typename TValue>
+void AppendValue(TNodeInfo & node, TValue const & value)
+{
+  // External-memory trie adds <string, value> pairs in a sorted
+  // order so the values are supposed to be accumulated in the
+  // sorted order and we can avoid sorting them before doing
+  // further operations such as ValueList construction.
+  using namespace std::rel_ops;
+  ASSERT(node.m_temporaryValueList.empty() || node.m_temporaryValueList.back() <= value, ());
```

Comment:
Please, write at least node.m_temporaryValueList size in a log message, because it will be hard to detect which part of disjunction caused invariant violation.